### PR TITLE
ENG-19131: Store encoding format in PBD

### PR DESCRIPTION
**Pull request by n3world**
Store the encoding format for a topic in the PBD. Add the encoding
format to PersistedMetadata and add it to the serdes.

### CODE REVIEW

File: src/frontend/org/voltdb/exportclient/PersistedMetadata.java [hunk 1]
```diff
@@ -28,27 +30,46 @@
...
             throw new IOException("Unsupported serialization version: " + version);
         }
-        return new PersistedMetadata(ExportRowSchema.deserialize(buf));
+        EncodeFormat format = EncodeFormat.byId(buf.get());
+        return new PersistedMetadata(format, ExportRowSchema.deserialize(buf));
     }
 
-    public PersistedMetadata(ExportRowSchema schema) {
+    public PersistedMetadata(Table table, int partitionId, long initialGenerationId, long generationId) {
+        m_format = table.getIstopic() ? EncodeFormat.checkedValueOf(table.getTopicformat()) : EncodeFormat.INVALID;
```
Comment (rdykiel) [inline src/frontend/org/voltdb/exportclient/PersistedMetadata.java]: Should it be EncodeFormat.VOLT?

File: src/frontend/org/voltdb/exportclient/PersistedMetadata.java [hunk 2]
```diff
@@ -28,27 +30,46 @@
...
+        m_schema = ExportRowSchema.create(table, partitionId, initialGenerationId, generationId);
+    }
+
+    private PersistedMetadata(EncodeFormat format, ExportRowSchema schema) {
```
Comment (rdykiel) [inline src/frontend/org/voltdb/exportclient/PersistedMetadata.java]: Should we use a builder pattern instead, in order to support future additions to the persisted metadata

File: src/frontend/org/voltdb/serdes/EncodeFormat.java [hunk 3]
```diff
@@ -37,18 +40,46 @@
      */
     public static EncodeFormat checkedValueOf(String name) {
         try {
-            return valueOf(EncodeFormat.class, name);
+            return valueOf(name);
         }
         catch (Exception ex) {
-            VoltDB.crashLocalVoltDB("Illegal encoding format " + name, true, ex);
+            throw VoltDB.crashLocalVoltDB("Illegal encoding format " + name, true, ex);
```
Comment (rdykiel) [inline src/frontend/org/voltdb/serdes/EncodeFormat.java]: You'll have to explain to me the rationale of using a throw here

File: src/frontend/org/voltdb/serdes/EncodeFormat.java [hunk 4]
```diff
@@ -25,8 +25,11 @@
  * An enum listing the encoding formats
  */
 public enum EncodeFormat {
-    CSV,
-    AVRO;
+    INVALID(-1),
+    CSV(0),
+    AVRO(1);
+
+    private final byte m_id;
```
Comment (rdykiel) [inline src/frontend/org/voltdb/serdes/EncodeFormat.java]: should it be "m_code" and a code() accessor like done in Errors? Also, comment why we use this.

File: src/frontend/org/voltdb/exportclient/PersistedMetadata.java (hunk 1 above)
Comment (n3world) [reply src/frontend/org/voltdb/exportclient/PersistedMetadata.java]: No this is the outbound format which is only valid for kipling. In export the outbound format is defined by the export client.

File: src/frontend/org/voltdb/serdes/EncodeFormat.java (hunk 3 above)
Comment (n3world) [reply src/frontend/org/voltdb/serdes/EncodeFormat.java]: You don't need a return. It tells java that execution stops here so nothing after this line will be executed

File: src/frontend/org/voltdb/exportclient/PersistedMetadata.java (hunk 2 above)
Comment (n3world) [reply src/frontend/org/voltdb/exportclient/PersistedMetadata.java]: If we need to support multiple different constructors to support the different types of PersistedMetadata then that could make sense but I think I'll defer that to when it is needed.

File: src/frontend/org/voltdb/serdes/EncodeFormat.java (hunk 4 above)
Comment (n3world) [reply src/frontend/org/voltdb/serdes/EncodeFormat.java]: I would say no. For Errors it is the error code so the word code makes sense. However in this case it is the identifier for the EncodeFormat so id makes more sense. I can add doc to this if you like

**Pull request merged**